ERCS/erc-9999.md

```diff
@@ -0,0 +1,195 @@
+        external
+        payable
+        returns (bytes[] memory results);
+
+    /// @dev This function is provided for frontends to detect support.
+    /// Currently, it returns 1.
+    function minimalBatchExecutorVersion() external pure returns (uint256);
+}
+```
+
+## Rationale
+
+### `minimalBatchExecutorVersion` for detection
+
+[ERC-165](./eip-165.md) is a very difficult standard to work with in due to inheritance issues (e.g. diamond inheritance).
+
+Additionally, we opt not to make the behavior of `execute` to return a magic number for a magic input for ease of implementation.
+
+### Only a single `execute` function
+
+There is no need to define a variant for the single call case. 
+
+Having only a single `execute` function greatly simplifies implementation and integration.
+
+It also reduces bytecode size and lessens the Solidity function dispatch load.
+
+### Use of `Call` struct instead of custom byte encoding decoding
+
+Most layer 2s already include calldata compression and forward the calldata savings to users.
+
+As such, we opt for simplicity to enable maximum compatibility and ease of implementation.
+
+Using a standard struct also helps with auditability, block explorer integrations, browser wallet integrations, and [EIP-712](./eip-712) signature generations.
+
+### Usage of a single `authData` bytes argument
+
+The single `authData` bytes argument also provides all the flexibility that is needed.
+
+## Backwards Compatibility
+
+No backwards compatibility issues.
+
+## Reference Implementation
+
+```solidity
+// SPDX-License-Identifier: CC0-1.0
+pragma solidity ^0.8.4;
+
+/// @notice Minimal batch executor mixin.
+/// @author Solady (https://github.com/vectorized/solady/blob/main/src/accounts/MinimalBatchExecutor.sol)
+abstract contract MinimalBatchExecutor {
+    /*´:°•.°+.*•´.*:˚.°*.˚•´.°:°•.°•.*•´.*:˚.°*.˚•´.°:°•.°+.*•´.*:*/
+    /*                          STRUCTS                           */
+    /*.•°:°.´+˚.*°.˚:*.´•*.+°.•°:´*.´•*.•°.•°:°.´:•˚°.*°.˚:*.´+°.•*/
+
+    /// @dev Call struct for the `execute` function.
+    struct Call {
+        address target;
+        uint256 value;
+        bytes data;
+    }
+
+    /*´:°•.°+.*•´.*:˚.°*.˚•´.°:°•.°•.*•´.*:˚.°*.˚•´.°:°•.°+.*•´.*:*/
+    /*                   FUNCTIONS TO OVERRIDE                    */
+    /*.•°:°.´+˚.*°.˚:*.´•*.+°.•°:´*.´•*.•°.•°:°.´:•˚°.*°.˚:*.´+°.•*/
+
+    /// @dev Ensures that `execute` can only be called by the correct caller or `authData`.
+    function _authorizeExecute(Call[] calldata calls, bytes calldata authData) internal virtual;
+
+    /*´:°•.°+.*•´.*:˚.°*.˚•´.°:°•.°•.*•´.*:˚.°*.˚•´.°:°•.°+.*•´.*:*/
+    /*                          EXECUTE                           */
+    /*.•°:°.´+˚.*°.˚:*.´•*.+°.•°:´*.´•*.•°.•°:°.´:•˚°.*°.˚:*.´+°.•*/
+
+    /// @dev Executes the `calls` and returns the results.
+    /// Reverts and bubbles up error if any call fails.
+    function execute(Call[] calldata calls, bytes calldata authData)
```

Comment:
consider using ERC7579 execute function here. 

>     function execute(bytes32 mode, bytes calldata executionCalldata) external;

would be a shame if wallets need to integrate another pattern.

the authData could be just appended bytes in `calldata` to perform the signature check, and ERC-4337 implementations could utilize the same function signature with access control `onlyEntryPoint` while handling the signature validation in `validateUserOp`

`minimalBatchExecutorVersion` could also be ERC-7579's `supportsExecutionMode`

Reply:
This standard will only support regular calls, in an atomic fashion. Delegatecalls won't be supported.

If users want to make multiple try transactions, they can just sign multiple transactions with parallel nonces.

I think it will not be too hard for frontends that have already supported 7579 to support this ERC. They will still need to craft the array of executions.

Side question: Do we actually need ERC-165 for ERC-7579? I would strongly prefer to never use ERC-165.